Project 4/Peihong_Karl_Proj_4_50375781.md

```diff
@@ -0,0 +1,1146 @@
+Note #2: The potential for this project to directly impact the ontology community is clear. SPARQL can be challenging, and there are few opportunities for drill practice like this. 
+Note #3: You will not be required to learn a programming language, though you will likely need to expand your comfort with computer science jargon; if you hit a wall, ask your peers for help; if the wall persists, ask me. 
+Note #4: Codewars provides a guidebook - https://docs.codewars.com/authoring/tutorials/create-first-kata/ - for creating problems; I strongly encourage you to read it, since the standard provided there is how I will be evaluating success. 
+```
+**Assignment Details**
+
+Problems on Codewars are ranked in terms of difficulty. The lowest "kata" - 8 - indicates a rather easy problem, while the highest kata - 1 - indicates a very challenging problem. 
+
+For our purposes, harder kata will be worth more points than easier kata, and you are required to submit enough kata to acquire 100 points according to the following point system: 
+
+  |   **kata**    |  **points**   |
+  | ------------- | ------------- |
+  |       1       |      35       |
+  |       2       |      25       |
+  |       3       |      20       |
+  |       4       |      10       |
+  |       5       |       5       |
+  |       6       |       3       |
+  |       7       |       2       |
+  |       8       |       0       |
+
+You're probably thinking, "why would I submit a level 8 kata if they're not worth any points?" Great question. Because everyone had to submit at least one level 8 kata. Otherwise, you're permitted to submit kata in any distribution you choose. For example, you might submit 2 problems for kata one (70 points), one for kata 3 (20 points), one for kata 4 (10 points), and one for kata 8 (0 points but required). 
+
+It is your responsibility and the responsibility of your peers reviewing your submission in PR to determine whether your submission is ranked appropriately. In the event that consensus is reached that your kata is ranked inappropriately, you must work with your peers to revise the submission so that it is either more or less challenging, accordingly. You are not permitted to submit new problems with different strengths after PRs are open, but must instead revise your PRs. So, think hard about how challenging your submission is. 
+
+There is one other option for those desiring a different sort of challenge. If you provide alongside your SPARQL submission a translation of the same problem into SQL, complete with documentations, solution, etc. then you may receive half points extra at that kata level (rounded up). For example, if you submit a SPARQL problem that is kata rank 1 and also submit a SQL version of that same problem, you  will receive 35+18=53 points. 
+
+****
```

Comment:
Dear Karl, 

Thank you for the invitation to review your Project #4. I thoroughly appreciate working with you because I always learn something new! I have added my comments and recommendations below. 

[1] Your kata lvl 8 is at least a kata lvl 7 according to the rubric. It looks good!

[2] Your kata lvl 4 is well formed! I do agree with you that this is a lvl 4 problem. Nice 3-step kata!

[3] Excellent SQL translation. It is a great way to ensure extra points! I should have considered doing this for my project. 

[4] Picky Boss Mark! Lol. It took me a while to see how you were getting this: "Compare them and tell me how many games with the most episodes have the worst selling performance." But I see where you compared both and how you ordered your data. This looks good! It is a tricky problem and I think you did a good job constructing this one! I agree that this one is kata lvl 4.

[5] Solid SQL translation!

[6] Mars Base Commander! Haha! This kata is one of my favorites. I think that because this one is logically demanding, it can be argued for as a kata lvl 2. Moreover, the SPARQL data construction for each rule is complicated. This is at least a kata lvl 3, but I could see this one being rated a lvl 2. 

I think if you are at 100, you can assume that it is a lvl 3 to be on the safe side. You can put in parentheses, however, that this may be a lvl 2 near the title. 

[7] Geralt of Rivia--another great kata with a well written story/description! I agree with you that this is a kata lvl 4. Nice job on this one.

[8] SQL version looks good! No comment!

[9] Wonderful kata, Karl. I really appreciate the diversity that you are expressing through the different kata you have written. I agree with you that the Confucius kata is a lvl 4. 

I have only one recommendation, it seems that there is a typo: "subsidie". Did you mean to write subsidize? 

[10] SQL Kongzi looks good!

[11] I really like the practicality of your Kata lvl 4 - Video Card Purchase. This is a well-formed kata. For this kata, I can see it being rated either a kata lvl 5 or kata lvl 4. I think it shades more toward kata lvl 4, however, I just wanted to share this point just so that you are aware of how I am receiving/categorizing it intuitively. I like the restriction that you made on limiting the reader from using ORDER BY and LIMIT. Your MIN solution works well.

[12] This is a good kata lvl 4! If you wanted to, you can easily make this one a kata lvl 3 by adding extra roles and relations in the data set and utilizing the math/aggregate functions. I agree that this kata is a lvl 4. 

In sum, your Project #4 was creative and had a diverse array of problems/solutions. According to my lights, there were only two kata that could be rated either higher or lower, i.e. Mars Base Commander and Video Card Purchase. However, it is unclear that it is worth changing their current ranking (they may end up canceling each other out if Mars Base Commander is a lvl 2 and Video Card Purchase is a lvl 5 lol).

Well done! Thanks again for allowing me to review your work.

Sincerely, 
Jaron

Reply:
@JaronJCheung Thank you for your nice review! Good catch for my typo "subsidie" and I have fixed it. Also, thank you for telling me your intuition about Video Card Purchase. I suspect I have two ways to keep on the safe side-- either by pretending Mars Base Commander to be level 2 or by adding some new well-crafted mathematical conditions to make Black Friday Deals as level 3. I am happy for any further progress on this.

Have a good sleep, dude!